src/extensions/metadata.rs

```diff
@@ -0,0 +1,166 @@
+//! Adds support for the IMAP ENABLE command specificed in [RFC
+//! 5464](https://tools.ietf.org/html/rfc5464).
+
+use crate::client::*;
+use crate::error::{Error, ParseError, Result};
+use crate::types::*;
+use imap_proto::types::{MailboxDatum, Metadata, Response};
+use std::io::{Read, Write};
+use std::sync::mpsc;
+
+trait CmdListItemFormat {
+    fn format_as_cmd_list_item(&self) -> String;
+}
+
+impl CmdListItemFormat for Metadata {
+    fn format_as_cmd_list_item(&self) -> String {
+        format!(
+            "{} {}",
+            validate_str(self.entry.as_str()).unwrap(),
+            self.value
+                .as_ref()
+                .map(|v| validate_str(v.as_str()).unwrap())
+                .unwrap_or("NIL".to_string())
+        )
+    }
+}
+
+/// Represents variants of DEPTH parameters for GETMETADATA command
+#[derive(Debug, Copy, Clone)]
+pub enum MetadataDepth {
+    /// Depth 0 for get metadata
+    Zero,
+    /// Depth 1 for get metadata
+    One,
+    /// Depth infinity for get metadata
+    Inf,
+}
+
+impl MetadataDepth {
+    fn depth_str<'a>(self) -> &'a str {
+        match self {
+            MetadataDepth::Zero => return "0",
+            MetadataDepth::One => return "1",
+            MetadataDepth::Inf => return "infinity",
+        }
+    }
+}
+
+fn parse_metadata<'a>(
+    mut lines: &'a [u8],
+    unsolicited: &'a mut mpsc::Sender<UnsolicitedResponse>,
+) -> Result<Vec<Metadata>> {
+    let mut res: Vec<Metadata> = Vec::new();
+    loop {
+        if lines.is_empty() {
+            break Ok(res);
+        }
+
+        match imap_proto::parse_response(lines) {
+            Ok((rest, resp)) => {
+                lines = rest;
+                match resp {
+                    Response::MailboxData(MailboxDatum::MetadataSolicited {
+                        mailbox: _,
+                        mut values,
+                    }) => {
+                        res.append(&mut values);
+                    }
+                    Response::MailboxData(MailboxDatum::MetadataUnsolicited {
+                        mailbox,
+                        values,
+                    }) => {
+                        unsolicited
+                            .send(UnsolicitedResponse::Metadata {
+                                mailbox: mailbox.to_string(),
+                                metadata_entries: values.iter().map(|s| s.to_string()).collect(),
+                            })
+                            .unwrap();
+                    }
+                    _ => {}
+                }
+            }
+            Err(_) => {
+                return Err(Error::Parse(ParseError::Invalid(lines.to_vec())));
+            }
+        }
+    }
+}
+
+/// Sends GETMETADATA command to the server and returns the list of entries and their values.
+pub fn get_metadata<'a, S: AsRef<str>, T: Read + Write>(
```

Comment:
This, and `set_metadata` should be methods _on_ `Session`, not free-standing methods like this. I'd also like to see the documentation a lot more extensive (again copying from the RFC). The docs for both methods should also make it very clear that this is an extension that may not be available for the given server. Ideally it'd also include some sample code for how to determine whether it's supported by checking the server capabilities.

Reply:
It should also document what each of the arguments do.